Examples/Io/Root/src/RootVertexPerformanceWriter.cpp

```diff
@@ -0,0 +1,280 @@
+
+int ActsExamples::RootVertexPerformanceWriter::
+    getNumberOfReconstructableVertices(
+        const SimParticleContainer& collection) const {
+  // map for finding frequency
+  std::map<int, int> fmap;
+
+  std::vector<int> reconstructableTruthVertices;
+
+  // traverse the array for frequency
+  for (const auto& p : collection) {
+    int secVtxId = p.particleId().vertexSecondary();
+    if (secVtxId != 0) {
+      // truthparticle from secondary vtx
+      continue;
+    }
+    int priVtxId = p.particleId().vertexPrimary();
+    fmap[priVtxId]++;
+  }
+
+  // iterate over the map
+  for (auto it : fmap) {
+    // Require at least 2 tracks
+    if (it.second > 1) {
+      reconstructableTruthVertices.push_back(it.first);
+    }
+  }
+
+  return reconstructableTruthVertices.size();
+}
+
+int ActsExamples::RootVertexPerformanceWriter::getNumberOfTruePriVertices(
+    const SimParticleContainer& collection) const {
+  // Vector to store indices of all primary vertices
+  std::set<int> allPriVtxIds;
+  for (const auto& p : collection) {
+    int priVtxId = p.particleId().vertexPrimary();
+    int secVtxId = p.particleId().vertexSecondary();
+    if (secVtxId != 0) {
+      // truthparticle from secondary vtx
+      continue;
+    }
+    // Insert to set, removing duplicates
+    allPriVtxIds.insert(priVtxId);
+  }
+  // Size of set corresponds to total number of primary vertices
+  return allPriVtxIds.size();
+}
+
+ActsExamples::ProcessCode ActsExamples::RootVertexPerformanceWriter::writeT(
+    const AlgorithmContext& ctx,
+    const std::vector<Acts::Vertex<Acts::BoundTrackParameters>>& vertices) {
+  // Exclusive access to the tree while writing
+  std::lock_guard<std::mutex> lock(m_writeMutex);
+
+  m_nrecoVtx = vertices.size();
+
+  ACTS_DEBUG("Number of reco vertices in event: " << m_nrecoVtx);
+  if (m_outputFile == nullptr)
+    return ProcessCode::SUCCESS;
+
+  // Read truth particle input collection
+  const auto& allTruthParticles =
+      ctx.eventStore.get<SimParticleContainer>(m_cfg.inputAllTruthParticles);
+  // Get number of true primary vertices
+  m_ntrueVtx = getNumberOfTruePriVertices(allTruthParticles);
+
+  ACTS_INFO("Total number of generated truth particles in event : "
+            << allTruthParticles.size());
+  ACTS_INFO(
+      "Total number of generated truth primary vertices : " << m_ntrueVtx);
+
+  // Read selected truth particle input collection
+  const auto& allSelectedParticles = ctx.eventStore.get<SimParticleContainer>(
+      m_cfg.inputSelectedTruthParticles);
+  // Get number of true primary vertices
+  m_nVtxDetAcceptance = getNumberOfTruePriVertices(allSelectedParticles);
+
+  ACTS_INFO("Total number of selected truth particles in event : "
+            << allSelectedParticles.size());
+  ACTS_INFO("Total number of detector-accepted truth primary vertices : "
+            << m_nVtxDetAcceptance);
+
+  // Read track-associated truth particle input collection
+  const auto& allAssociatedTruthParticles =
```

Comment:
And how about `associatedTruthParticles`?

Reply:
Done.